File: Chapter 09/C#/Section_9_2_3_Stateful_Vs_Stateless_Design/Infrastructure/SimulatedCloudStorageProvider.cs
using System.Collections.Generic;
using System.IO;
using Chapter09.StatefulVsStateless.Services;

namespace Chapter09.StatefulVsStateless.Infrastructure;

/// <summary>
/// THE STATELESS ADAPTER (Cloud Native Design):
/// 
/// TEACHING NOTE:
/// This adapter moves the "state" out of the individual web servers and into 
/// a centralized, external location (simulating Amazon S3). 
/// 
/// Because our UserService depends entirely on the IStorageProvider interface, 
/// we can swap from the fragile LocalStorageProvider to this robust Cloud provider 
/// without changing a single line of our business logic!
/// </summary>
public class SimulatedCloudStorageProvider : IStorageProvider
{
    // The static dictionary acts as our "External Cloud". 
    // Even if we instantiate 100 web servers, they all point to this exact same data store.
    private static readonly Dictionary<string, string> _s3BucketSimulator = new();
    private readonly string _bucketName;

    public SimulatedCloudStorageProvider(string bucketName)
    {
        _bucketName = bucketName;
    }

    public void Save(string fileName, string data)
    {
        string s3Key = $"{_bucketName}/{fileName}";
        _s3BucketSimulator[s3Key] = data;
    }

    public string Get(string fileName)
    {
        string s3Key = $"{_bucketName}/{fileName}";
        if (!_s3BucketSimulator.ContainsKey(s3Key))
        {
            throw new FileNotFoundException($"404 Not Found in S3 Bucket: {s3Key}");
        }
        return _s3BucketSimulator[s3Key];
    }
}

File: Chapter 09/JS/section_9_2_3_stateful_vs_stateless_design/infrastructure/simulatedCloudStorageProvider.js
import { of, throwError } from 'rxjs';

/**
 * THE STATELESS ADAPTER (Cloud Native Design):
 * * TEACHING NOTE:
 * This adapter moves the "state" out of the individual web servers and into 
 * a centralized, external location (simulating Amazon S3). 
 * * Because our UserService relies on Dependency Injection, we can swap 
 * from the fragile LocalStorage to this robust Cloud provider without 
 * changing a single line of our business logic!
 * * We use the `of()` operator from RxJS to create an Observable that 
 * instantly emits our simulated S3 data and then completes the stream.
 */

// This Map acts as our "External Cloud". Because it lives outside the class instance,
// any server that instantiates this provider will access the exact same data store.
const s3BucketSimulator = new Map();

export class SimulatedCloudStorageProvider {
    
    /**
     * @param {string} bucketName - The name of the central cloud bucket.
     */
    constructor(bucketName) {
        this.bucketName = bucketName;
        // In real life, you would initialize the AWS SDK S3 client here.
    }

    /**
     * Saves data to the centralized cloud bucket.
     * @param {string} fileName 
     * @param {string} data 
     * @returns {import('rxjs').Observable<void>}
     */
    save(fileName, data) {
        const s3Key = `${this.bucketName}/${fileName}`;
        s3BucketSimulator.set(s3Key, data);
        
        // Return an empty observable that completes immediately
        return of(void 0); 
    }

    /**
     * Retrieves data from the centralized cloud bucket.
     * @param {string} fileName 
     * @returns {import('rxjs').Observable<string>}
     */
    get(fileName) {
        const s3Key = `${this.bucketName}/${fileName}`;
        
        if (!s3BucketSimulator.has(s3Key)) {
            return throwError(() => new Error(`404 Not Found in S3 Bucket: ${s3Key}`));
        }
        
        return of(s3BucketSimulator.get(s3Key));
    }
}